### DISCUSSION TITLE
Why calling a lambda in expressions causes an error

### DISCUSSION THREAD

**Discussion by xbyte3** [wiremod/wire#2886] [Q&A] [answered]
Code:
```
const A = function() {
    return 1
}
B = A() # <- Error: Cannot use void in expression position
```
Why calling a lambda in expressions causes an error "Cannot use void in expression position"? I think this is a bug.

**Comment by thegrb93** [comment]
I'm guessing it's because your function doesn't have a return type?

**Reply by vurvdev** [reply]
Nah lambdas (and functions now, too) have implicit return types, so the compiler will recognize it based on what type you return

**Comment by Denneisk** [comment]
Lambda calls require using a type designator. Just how it was made.

**Comment by vurvdev** [comment, marked as answer]
Lambdas are typed erased to just being a single `function` type. That's how tables and arrays work too.

So whenever you call a lambda, the compiler needs to know what return type you're expecting, to check the signature (stored at runtime).

This comes from
1. String calls working the same way
2. Very strict / generic typing being very foreign to E2 (dunno if the leap will ever be made in E2, might be a reason for E3 to happen)
3. Type system / extension system would need to be changed (potentially not that much, could just store fn signatures as `fn(a,b,c):d`)
